p2p/src/codec.rs

```diff
@@ -0,0 +1,244 @@
+				if let Err(e) = self.stream.read_exact(&mut self.buffer[pre_len..]) {
+					// Undo reserved bytes on a failed read
+					self.buffer.truncate(pre_len);
+					return Err(e.into());
+				}
+				self.bytes_read += to_read;
+			}
+			match &mut self.state {
+				None => {
+					// Parse header and keep reading
+					let mut raw = self.buffer.split_to(next_len).freeze();
+					let mut reader = BufReader::new(&mut raw, self.version);
+					let header = MsgHeaderWrapper::read(&mut reader)?;
+					self.state = Header(header);
+				}
+				Header(Known(header)) => {
+					let mut raw = self.buffer.split_to(next_len).freeze();
+					if header.msg_type == Type::Headers {
+						// Special consideration for a list of headers, as we want to verify and process
+						// them as they come in instead of only after the full list has been received
+						let mut reader = BufReader::new(&mut raw, self.version);
+						let items_left = reader.read_u16()? as usize;
+						self.state = BlockHeaders {
+							bytes_left: header.msg_len as usize - 2,
+							items_left,
+							headers: Vec::with_capacity(min(HEADER_BATCH_SIZE, items_left)),
+						};
+					} else {
+						// Return full message
+						let msg = decode_message(header, &mut raw, self.version);
+						self.state = None;
+						return msg;
+					}
+				}
+				Header(Unknown(_, msg_type)) => {
+					// Discard body and return
+					let msg_type = *msg_type;
+					self.buffer.advance(next_len);
+					self.state = None;
+					return Ok(Message::Unknown(msg_type));
+				}
+				BlockHeaders {
+					bytes_left,
+					items_left,
+					headers,
+				} => {
+					if *bytes_left == 0 {
+						// Incorrect item count
+						self.state = None;
+						return Err(Error::BadMessage);
+					}
+
+					let mut reader = BufReader::new(&mut self.buffer, self.version);
+					let header: UntrustedBlockHeader = reader.body()?;
+					let bytes_read = reader.bytes_read() as usize;
+					headers.push(header.into());
+					*bytes_left = bytes_left.saturating_sub(bytes_read);
+					*items_left -= 1;
+
+					if headers.len() == HEADER_BATCH_SIZE || *items_left == 0 {
+						let mut h = Vec::with_capacity(min(HEADER_BATCH_SIZE, *items_left));
+						mem::swap(headers, &mut h);
+						if *items_left == 0 {
+							let bytes_left = *bytes_left;
+							self.state = None;
+							if bytes_left > 0 {
+								return Err(Error::BadMessage);
+							}
+						}
+						return Ok(Message::Headers(h));
+					}
+				}
+				Attachment(left, meta, now) => {
+					let raw = self.buffer.split_to(next_len).freeze();
+					*left -= next_len;
+					if now.elapsed().as_secs() > 10 {
+						*now = Instant::now();
+						debug!("attachment: {}/{}", meta.size - *left, meta.size);
+					}
+					let update = AttachmentUpdate {
+						read: next_len,
+						left: *left,
+						meta: Arc::clone(meta),
+					};
+					if *left == 0 {
+						self.state = None;
+						debug!("attachment: DONE");
+					}
+					return Ok(Message::Attachment(update, Some(raw)));
```

Comment:
Does `raw` here represent _everything_ read for this attachment so far? Or just the part read in this particular call to `read_inner`?

Reply:
`raw` is a chunk of 48000 bytes. Storing the full attachment in a buffer would be very inefficient, as we would have to allocate a lot of memory.

Reply:
I see it now. And then we write these bytes to the file itself in the main `reader_thread` loop.